guides/github.md

```diff
@@ -0,0 +1,130 @@
+dependencies. Both of these tools can be configured to send a PR at regular
+intervals (e.g., once a week).
+
+It is recommended to install [Scorecards action][scorecard] for public
+repositories and trying to improve the score as high as possible. This is a
+scanner for security best practices, most of which are already discussed in
+this document.
+
+## Securing GitHub Actions
+
+It is preferable to use the hosted runners as possible. The large runners
+supported by GitHub come with usage costs, but are better from the point of
+view of supply chain security than hosting your own infrastructure for CI.
+
+All GitHub Actions workflows should restrict [permissions][gha-permissions] to
+the minimum scope needed. Scope the permissions at job level, instead of at
+workflow level.
+
+```
+  check-tests:
+    permissions:
+      actions: read
+```
+
+Minimize usage of actions that create PRs or push code to branch. Thoroughly
+inspect actions that can approve PRs and workflows that are triggered after a
+PR has been approved (term-of-check-vs-term-of-use type of concerns).
+
+If using actions outside of your organisation, these should be pinned by
+commit hash. Don't pin be version tag as these tags can be forced pushed. The
+[Scorecards Pinned Dependencies workflow][scorecard] can help identify which
+actions should still be pinned. [Dependabot][dependabot] is able to upgrade
+the versions of these actions periodically:
+
+```
+      - name: Upload sdist as an artifact for later jobs in workflow
+        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        with:
+          ...
+```
+
+If secrets are used as arguments to commands for the workflows, pass them via
+environment variables. The action log can leak command line arguments!
+
+If using secrets, these should be scrubbed as soon as they are no longer
+needed. Make sure they don't show up in logs, including debug runs of the
+workflow!
+
+Be extra careful of GitHub context variables (e.g.,
+`github.event.issue.title`) which are vulnerable to command injection attacks.
+Don't use these directly in commands. Prefer passing them via environment
+variables. Use [Scorecards][scorecard] to identify dangerous workflows.
+
+Audit carefully all workflows of type `pull_request_target` and
+`workflow_run`. For these, don't use secrets at all and separate the sensitive
+parts into separate runs. Validate all inputs and only run these workflows on
+protected branches. Never use `pull_request_head.ref` as the branch head
+reference since this can be updated between the time the PR gets approved and
+the time the CI workflow starts executing.
+
+### Using self-hosted runners
+
+If using self-hosted runners, create 2 separate pools: one pool to be used
+only for CI for PRs and another one to be used for releases and other
+sensitive jobs. Ideally, don't share these pools across multiple projects in
+the same organisation.
```

Comment:
What is the reasoning? Github can be configured to require approval before workflows are run from outside contributors. This attack vector is already covered.

Reply:
Sure, if you approve all PRs before running these actions then the vector is covered. But many people consider that asking for review before any CI runs for everyone is too much. So they either pick the middle ground and require approval only for the first contribution (which then is bypassed by a typo fix) or add the contributor as a collaborator to the project.

Plus, if you missed something during review, or the attacker manages to exploit a TOCTOU issue, they could still poison the runners.

Reply:
It is mentioned in the first warning box in the doc: https://github.com/haskell/security-advisories/blob/main/guides/github.md